LedgerCloseTest#testLedgerCloseWithConsistentLength failed with IllegalReferenceCountException

### Description
When running LedgerCloseTest, the `testLedgerCloseWithConsistentLength` failed with IllegalReferenceCountException
```
2023-02-24T16:22:04.4265552Z [ERROR] Tests run: 4, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 26.555 s <<< FAILURE! - in org.apache.bookkeeper.client.LedgerCloseTest
2023-02-24T16:22:04.4269548Z [ERROR] org.apache.bookkeeper.client.LedgerCloseTest.testLedgerCloseWithConsistentLength  Time elapsed: 2.89 s  <<< ERROR!
2023-02-24T16:22:04.4270352Z io.netty.util.IllegalReferenceCountException: refCnt: 0, increment: 1
2023-02-24T16:22:04.4270960Z 	at io.netty.util.internal.ReferenceCountUpdater.retain0(ReferenceCountUpdater.java:132)
2023-02-24T16:22:04.4271598Z 	at io.netty.util.internal.ReferenceCountUpdater.retain(ReferenceCountUpdater.java:119)
2023-02-24T16:22:04.4272204Z 	at io.netty.util.AbstractReferenceCounted.retain(AbstractReferenceCounted.java:61)
2023-02-24T16:22:04.4273013Z 	at org.apache.bookkeeper.util.ByteBufList.retain(ByteBufList.java:281)
2023-02-24T16:22:04.4273614Z 	at org.apache.bookkeeper.proto.BookieClientImpl.addEntry(BookieClientImpl.java:306)
2023-02-24T16:22:04.4274246Z 	at org.apache.bookkeeper.client.PendingAddOp.sendWriteRequest(PendingAddOp.java:148)
2023-02-24T16:22:04.4274850Z 	at org.apache.bookkeeper.client.PendingAddOp.initiate(PendingAddOp.java:263)
2023-02-24T16:22:04.4275451Z 	at org.apache.bookkeeper.client.LedgerHandle.doAsyncAddEntry(LedgerHandle.java:1358)
2023-02-24T16:22:04.4276060Z 	at org.apache.bookkeeper.client.LedgerHandle.asyncAddEntry(LedgerHandle.java:1056)
2023-02-24T16:22:04.4276673Z 	at org.apache.bookkeeper.client.LedgerHandle.asyncAddEntry(LedgerHandle.java:1051)
2023-02-24T16:22:04.4277407Z 	at org.apache.bookkeeper.client.LedgerHandle.asyncAddEntry(LedgerHandle.java:1003)
2023-02-24T16:22:04.4278133Z 	at org.apache.bookkeeper.client.LedgerCloseTest.testLedgerCloseWithConsistentLength(LedgerCloseTest.java:88)
2023-02-24T16:22:04.4278851Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-02-24T16:22:04.4279492Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2023-02-24T16:22:04.4280187Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-02-24T16:22:04.4280778Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2023-02-24T16:22:04.4281289Z 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
2023-02-24T16:22:04.4281871Z 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2023-02-24T16:22:04.4282496Z 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
2023-02-24T16:22:04.4283121Z 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2023-02-24T16:22:04.4283729Z 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2023-02-24T16:22:04.4284326Z 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2023-02-24T16:22:04.4284945Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
2023-02-24T16:22:04.4285587Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
2023-02-24T16:22:04.4286148Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2023-02-24T16:22:04.4286586Z 	at java.base/java.lang.Thread.run(Thread.java:829)
```
Failed CI: https://github.com/apache/bookkeeper/actions/runs/4263713315/jobs/7421980838

We can reproduce it by running LedgerCloseTest all tests instead of running the `testLedgerCloseWithConsistentLength` individual test.

After I reverted https://github.com/apache/bookkeeper/pull/3784, the test passed.

--- a/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java
+++ b/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java
@@ -1870,7 +1870,8 @@ public class LedgerHandle implements WriteHandle {
                 LOG.debug("Ensemble change is disabled. Retry sending to failed bookies {} for ledger {}.",
                     failedBookies, ledgerId);
             }
-            unsetSuccessAndSendWriteRequest(getCurrentEnsemble(), failedBookies.keySet());
+            executeOrdered(() ->
+                    unsetSuccessAndSendWriteRequest(getCurrentEnsemble(), failedBookies.keySet()));
             return;
         }
 

--- a/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java
+++ b/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java
@@ -231,7 +231,7 @@ class PendingAddOp implements WriteCallback {
     /**
      * Initiate the add operation.
      */
-    public void initiate() {
+    public synchronized void initiate() {
         hasRun = true;
         if (callbackTriggered) {
             // this should only be true if the request was failed due
@@ -452,7 +452,7 @@ class PendingAddOp implements WriteCallback {
     }
 
 
-    private void maybeRecycle() {
+    private synchronized void maybeRecycle() {
         /**
          * We have opportunity to recycle two objects here.
          * PendingAddOp#toSend and LedgerHandle#pendingAddOp
@@ -482,7 +482,7 @@ class PendingAddOp implements WriteCallback {
         }
     }
 
-    public void recyclePendAddOpObject() {
+    public synchronized void recyclePendAddOpObject() {
         entryId = LedgerHandle.INVALID_ENTRY_ID;
         currentLedgerLength = -1;
         if (payload != null) {

--- a/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java
+++ b/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java
@@ -358,7 +358,7 @@ class PendingAddOp implements WriteCallback {
             } else {
                 LOG.warn("Failed to write entry ({}, {}) to bookie ({}, {}): {}",
                         ledgerId, entryId, bookieIndex, addr, BKException.getMessage(rc));
-                lh.executeOrdered(() -> lh.handleBookieFailure(ImmutableMap.of(bookieIndex, addr)));
+                lh.handleBookieFailure(ImmutableMap.of(bookieIndex, addr));
             }
             return;
         }

--- a/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java
+++ b/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java
@@ -358,7 +358,7 @@ class PendingAddOp implements WriteCallback {
             } else {
                 LOG.warn("Failed to write entry ({}, {}) to bookie ({}, {}): {}",
                         ledgerId, entryId, bookieIndex, addr, BKException.getMessage(rc));
-                lh.handleBookieFailure(ImmutableMap.of(bookieIndex, addr));
+                lh.executeOrdered(() -> lh.handleBookieFailure(ImmutableMap.of(bookieIndex, addr)));
             }
             return;
         }

